diff --git a/src/exports/node/index.js b/src/exports/node/index.js
index a69b7ae6d906745f9a3a6e2b00f5e7134f4fee77..63aa6dbd0ec8146da6cdc053b65d47f13657baa6 100644
--- a/src/exports/node/index.js
+++ b/src/exports/node/index.js
@@ -106,6 +106,10 @@ function get_raw_body(req, body_size_limit) {
 // TODO 3.0 make the signature synchronous?
 // eslint-disable-next-line @typescript-eslint/require-await
 export async function getRequest({ request, base, bodySizeLimit }) {
+	if(Symbol.for('::bunternal::') in request.socket) {
+		return request.socket[Symbol.for('::bunternal::')][2];
+	}
+
 	let headers = /** @type {Record<string, string>} */ (request.headers);
 	if (request.httpVersionMajor >= 2) {
 		// the Request constructor rejects headers with ':' in the name
@@ -140,6 +144,11 @@ export async function getRequest({ request, base, bodySizeLimit }) {
 // TODO 3.0 make the signature synchronous?
 // eslint-disable-next-line @typescript-eslint/require-await
 export async function setResponse(res, response) {
+	if('Bun' in globalThis && '_reply' in res) {
+		res._reply(response);
+		return;
+	}
+
 	for (const [key, value] of response.headers) {
 		try {
 			res.setHeader(
